FlowCrypt/Functionality/DataManager/Encrypted Storage/KeyChainService.swift

```diff
@@ -19,67 +19,85 @@ actor KeyChainService {
-            // and that would corrupt existing storage
-            // https://github.com/FlowCrypt/flowcrypt-ios/issues/1373
-            throw AppErr.general("KeyChainService: protected data is not available")
-        }
-        let dynamicPartIndex = "indexSecureKeychainPrefix"
-        if let storedDynamicPart = UserDefaults.standard.string(forKey: dynamicPartIndex) {
-            return constructKeychainIndex(dynamicPart: storedDynamicPart)
-        }
-        let newDynamicPart = try newRandomString()
-        UserDefaults.standard.set(newDynamicPart, forKey: dynamicPartIndex)
-        return constructKeychainIndex(dynamicPart: newDynamicPart)
+    private enum Constants {
+        static let legacyKeychainIndexPrefixInUserDefaults = "indexSecureKeychainPrefix"
+        static let keychainPropertyKey = "app-storage-encryption-key"
     }
 
-    @MainActor private func newRandomString() throws -> String {
-        logger.logInfo("newRandomString - generating new KeyChain index")
-        guard let randomBytes = CoreHost().getSecureRandomByteNumberArray(12) else {
-            throw AppErr.general("KeyChainService.newRandomString - randomBytes are nil")
+    @MainActor var storageEncryptionKey: Data {
+        get throws {
+            if let key = try fetchEncryptionKey() {
+                return key
+            }
+
+            // TODO: Should be removed after all users migrated to new keychain
+            if let legacyKey = try fetchLegacyEncryptionKey() {
+                try saveStorageEncryptionKey(data: legacyKey)
+                return legacyKey
+            }
+
+            return try generateStorageEncryptionKey()
         }
-        return Data(randomBytes)
-            .base64EncodedString()
-            .replacingOccurrences(of: "[^A-Za-z0-9]+", with: "", options: [.regularExpression])
     }
 
-    @MainActor private func constructKeychainIndex(dynamicPart: String) -> String {
-        return dynamicPart + "-indexStorageEncryptionKey"
+    @MainActor private func fetchLegacyEncryptionKey() throws -> Data? {
+        let userDefaults = UserDefaults.standard
+        let prefixKey = Constants.legacyKeychainIndexPrefixInUserDefaults
+
+        guard let storedDynamicPart = userDefaults.string(forKey: prefixKey) else { return nil }
+
+        userDefaults.removeObject(forKey: prefixKey)
+        let storageKey = storedDynamicPart + "-indexStorageEncryptionKey"
+
+        guard let encryptionKey = try fetchEncryptionKey(property: storageKey) else {
+            if try !EncryptedStorage.doesStorageFileExist {
+                throw AppErr.general("KeyChainService: failed to get encryption key prefix")
+            }
+            return nil
+        }
+
+        return encryptionKey
     }
 
-    @MainActor private func generateAndSaveStorageEncryptionKey() throws {
-        logger.logInfo("generateAndSaveStorageEncryptionKey")
+    @MainActor private func generateStorageEncryptionKey() throws -> Data {
```

Comment:
As currently implemented, this method should still be called `generateAndSaveStorageEncryptionKey` - since calling the method will do both actions as currently implemented.

If you'd like to have two distinct methods called `generateStorageEncryptionKey` and `saveStorageEncryptionKey`, then you should not have one of them call the other, instead call both independently.

A method name should describe all important things that will happen when calling it, when possible, to reduce surprises.

Reply:
Yeah, you're right - I forgot to rename it back after adding migration code.
Currently `saveStorageEncryptionKey` also used when migrating from `UserDefaults`, I'll move it's code back to `generateAndSaveStorageEncryptionKey` after removing migration functionality.